File: webapps/src/main/java/org/camunda/bpm/webapp/impl/security/auth/AuthenticationService.java
/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding copyright
 * ownership. Camunda licenses this file to you under the Apache License,
 * Version 2.0; you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.camunda.bpm.webapp.impl.security.auth;

import static org.camunda.bpm.engine.authorization.Permissions.ACCESS;
import static org.camunda.bpm.engine.authorization.Resources.APPLICATION;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;

import javax.ws.rs.core.Response.Status;

import org.camunda.bpm.engine.AuthorizationService;
import org.camunda.bpm.engine.ProcessEngine;
import org.camunda.bpm.engine.identity.Group;
import org.camunda.bpm.engine.identity.Tenant;
import org.camunda.bpm.engine.identity.User;
import org.camunda.bpm.engine.rest.exception.InvalidRequestException;
import org.camunda.bpm.webapp.impl.util.ProcessEngineUtil;

public class AuthenticationService {

  public static final String[] APPS = new String[] { "cockpit", "tasklist", "admin"};
  public static final String APP_WELCOME = "welcome";

  public Authentication createAuthenticate(String engineName, String username) {
    return createAuthenticate(engineName, username, null, null);
  }

  public Authentication createAuthenticate(String engineName, String username, List<String> groupIds, List<String> tenantIds) {
    ProcessEngine processEngine = ProcessEngineUtil.lookupProcessEngine(engineName);

    if(processEngine == null) {
      throw new InvalidRequestException(Status.BAD_REQUEST, "Process engine with name "+engineName+" does not exist");
    }

    return createAuthenticate(processEngine, username, groupIds, tenantIds);
  }

  public Authentication createAuthenticate(ProcessEngine processEngine, String username, List<String> groupIds, List<String> tenantIds) {

    String userId = username;

    User user = processEngine.getIdentityService()
        .createUserQuery()
        .userId(username)
        .singleResult();

    if (user != null && user.getId() != null && !user.getId().isEmpty()) {
      userId = user.getId();
    }

    // make sure authentication is executed without authentication :)
    processEngine.getIdentityService().clearAuthentication();

    if (groupIds == null) {
      groupIds = getGroupsOfUser(processEngine, userId);
    }

    if (tenantIds == null) {
      tenantIds = getTenantsOfUser(processEngine, userId);
    }

    // check user's app authorizations
    AuthorizationService authorizationService = processEngine.getAuthorizationService();

    HashSet<String> authorizedApps = new HashSet<>();
    authorizedApps.add(APP_WELCOME);

    if (processEngine.getProcessEngineConfiguration().isAuthorizationEnabled()) {
      for (String application: APPS) {
        if (isAuthorizedForApp(authorizationService, userId, groupIds, application)) {
          authorizedApps.add(application);
        }
      }

    } else {
      Collections.addAll(authorizedApps, APPS);
    }

    // create new authentication
    UserAuthentication newAuthentication = new UserAuthentication(userId, processEngine.getName());
    newAuthentication.setGroupIds(groupIds);
    newAuthentication.setTenantIds(tenantIds);
    newAuthentication.setAuthorizedApps(authorizedApps);

    return newAuthentication;
  }

  public List<String> getTenantsOfUser(ProcessEngine engine, String userId) {
    List<Tenant> tenants = engine.getIdentityService().createTenantQuery()
      .userMember(userId)
      .includingGroupsOfUser(true)
      .list();

    List<String> tenantIds = new ArrayList<>();
    for(Tenant tenant : tenants) {
      if (tenant != null && tenant.getId() != null)
      {
        tenantIds.add(tenant.getId());
      }
    }
    return tenantIds;
  }

  public List<String> getGroupsOfUser(ProcessEngine engine, String userId) {
    List<Group> groups = engine.getIdentityService().createGroupQuery()
      .groupMember(userId)
      .list();

    List<String> groupIds = new ArrayList<>();
    for (Group group : groups) {
      if (group != null && group.getId() != null)
      {
        groupIds.add(group.getId());
      }
    }
    return groupIds;
  }

  protected boolean isAuthorizedForApp(AuthorizationService authorizationService, String username, List<String> groupIds, String application) {
    return authorizationService.isUserAuthorized(username, groupIds, ACCESS, APPLICATION, application);
  }

}

File: webapps/frontend/camunda-commons-ui/lib/auth/service/authenticationService.js
/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding copyright
 * ownership. Camunda licenses this file to you under the Apache License,
 * Version 2.0; you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

require('camunda-bpm-sdk-js/vendor/angular');
var $ = require('jquery');
var Authentication = require('../util/authentication');

/**
 * An authentication provider
 * @name AuthenticationService
 * @memberof cam.common.auth.service
 * @type angular.service
 */
module.exports = [
  '$rootScope',
  '$q',
  '$http',
  'Uri',
  '$route',
  function($rootScope, $q, $http, Uri, $route) {
    function emit(event, a, b) {
      $rootScope.$broadcast(event, a, b);
    }

    function parse(response) {
      if (response.status !== 200) {
        return $q.reject(response);
      }

      var data = response.data;

      return new Authentication({
        name: data.userId,
        authorizedApps: data.authorizedApps
      });
    }

    function update(authentication) {
      $rootScope.authentication = authentication;
      emit('authentication.changed', authentication);
    }

    /**
     * Force an update of the current application authentication
     *
     * @method AuthenticationService#update
     *
     * @param {Authentication} authentication
     */
    this.updateAuthentication = update;

    //////// login logic ///////////////

    /**
     * Login into the application with the given credentials
     *
     * @method AuthenticationService#login
     *
     * @param  {String} username
     * @param  {String} password
     *
     * @return {Promise<Authentication, Response>} the promise with the new authentication or failure response
     */
    this.login = function(username, password) {
      var form = $.param({
        username: username,
        password: password
      });

      function success(authentication) {
        // When the next subsequent request after the location change is
        // a modifying request (POST), no valid CSRF token is available
        $http.get(Uri.appUri('engine://engine/')).then(function() {
          update(authentication);
          emit('authentication.login.success', authentication);
        });

        return authentication;
      }

      function error(response) {
        emit('authentication.login.failure', response);
        return $q.reject(response);
      }

      function performRequest() {
        return $http({
          method: 'POST',
          url: Uri.appUri('admin://auth/user/:engine/login/:appName'),
          data: form,
          headers: {
            'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8'
          }
        });
      }

      // We have to perform a GET request before login to make sure we have an up to date CSRF-Cookie
      return $http
        .get(Uri.appUri('engine://engine/'))
        .then(performRequest)
        .then(parse)
        .then(success, error);
    };

    //////// logout logic ///////////////

    /**
     * Logout from the application.
     *
     * @method AuthenticationService#logout
     *
     * @return {Promise<Void, Response>} a promise that resolves to the failed response on error
     */
    this.logout = function(engine) {
      function success(response) {
        update(null);
        emit('authentication.logout.success', response);
      }

      function error(response) {
        emit('authentication.logout.failure', response);
        return $q.reject(response);
      }

      return $http
        .post(
          Uri.appUri('admin://auth/user/' + (engine || ':engine') + '/logout')
        )
        .then(success, error);
    };

    //////// get current auth ///////////////

    /**
     * A promise that returns the current authentication, either fetched from server
     * or from local application context. Will be reset when the authentication changed only.
     */
    var authenticationPromise;

    $rootScope.$on('authentication.changed', function(e, authentication) {
      authenticationPromise = $q[authentication ? 'when' : 'reject'](
        authentication
      );

      if (!authentication) {
        $route.reload();
      }
    });

    /**
     * Resolves the current authentication, either fetched from server or locally.
     *
     * @method AuthenticationService#getAuthentication
     *
     * @return {Promise<Authentication, Response>} a promise resolving to the current users authentication
     */
    this.getAuthentication = function() {
      function success(authentication) {
        update(authentication);
        return authentication;
      }

      if (!authenticationPromise) {
        // if already authenticated, return stored credentials
        if ($rootScope.authentication) {
          authenticationPromise = $q.when($rootScope.authentication);
        } else {
          authenticationPromise = $http
            .get(Uri.appUri('admin://auth/user/:engine'))
            .then(parse)
            .then(success);
        }
      }

      return authenticationPromise;
    };

    //////// secure route integration ///////////////

    $rootScope.$on('$routeChangeStart', function(event, next) {
      // secures routes that are annotated with
      //
      // {
      //   authentication: 'required'
      // }

      if (next.authentication) {
        if (!next.resolve) {
          next.resolve = {};
        }

        if (!next.resolve.authentication) {
          next.resolve.authentication = [
            'AuthenticationService',
            function(AuthenticationService) {
              return AuthenticationService.getAuthentication().catch(function(
                response
              ) {
                if (next.authentication === 'optional') {
                  return null;
                } else {
                  emit('authentication.login.required', next);
                  return $q.reject(response);
                }
              });
            }
          ];
        }
      }
    });
  }
];
